File: leetcode/740. Delete and Earn.cpp
// 740. 删除并获得点数
#include <algorithm>
#include <cmath>
#include <vector>
using namespace std;
class Solution
{
public:
    int deleteAndEarn(vector<int> &nums)
    {
        int n = nums.size();
        int siz_e = *max_element(nums.begin(), nums.end());
        vector<int> sum(siz_e + 1, 0);
        for (auto elem : nums)
        {
            sum[elem] += elem;
        }
        vector<int> dp(sum.size(), 0);
        dp[0] = sum[0], dp[1] = max(sum[0], sum[1]);
        for (int i = 2; i < sum.size(); i++)
        {
            dp[i] = max(dp[i - 2] + sum[i], dp[i - 1]);
        }
        return dp[sum.size() - 1];
    }
};

File: leetcode/otherLanguage/740. Delete and Earn.java
// 740. 删除并获得点数
class Solution
{
public
    int deleteAndEarn(int[] nums)
    {
        if (nums == null || nums.length == 0)
        {
            return 0;
        }
        else if (nums.length == 1)
        {
            return nums[0];
        }
        int len = nums.length;
        int max = nums[0];
        for (int i = 0; i < len; ++i)
        {
            max = Math.max(max, nums[i]);
        }
        //      构造一个新的数组all
        int[] all = new int[max + 1];
        for (int item : nums)
        {
            all[item]++;
        }
        int[] dp = new int[max + 1];
        dp[1] = all[1] * 1;
        dp[2] = Math.max(dp[1], all[2] * 2);
        //      动态规划求解
        for (int i = 2; i <= max; ++i)
        {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + i * all[i]);
        }
        return dp[max];
    }
}